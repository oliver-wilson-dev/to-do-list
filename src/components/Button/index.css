@import "../../styles/variables.css";

.button {
    padding: 10px 40px;
    border: 3px solid #f0134d;
    box-sizing: border-box;
    text-transform:uppercase;
    font-weight:400;
    text-align:center;
    transition: all 0.15s;
    font-size: 18px;
    background-color: var(--primaryBackgroundColour);
}
