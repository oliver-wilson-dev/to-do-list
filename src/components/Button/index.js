import React from 'react';
import PropTypes from 'prop-types';
import styles from './index.css';

const Button = ({ incrementCount }) => (
  <button className={styles.button} onClick={incrementCount} type="button">Click me!</button>
);

Button.propTypes = {
  incrementCount: PropTypes.func.isRequired
};

export default Button;
