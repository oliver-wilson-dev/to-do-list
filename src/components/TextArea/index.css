.textArea {
    resize: none;
    overflow: auto;
    background-color: var(--primaryBackgroundColour);
    font-size: inherit;
    padding: 5px 10px;
    border: 3px solid var(--primaryTextColor);
    margin-right: 5px;
}