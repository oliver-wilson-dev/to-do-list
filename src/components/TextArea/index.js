import React from 'react';
import PropTypes from 'prop-types';
import styles from './index.css';

const TextArea = ({
  onChange, value, additionalStyles, reference
}) => {
  React.useEffect(() => {
    reference.current.focus();
  }, []);

  return (
    <textarea
      className={`${styles.textArea} ${additionalStyles || ''}`}
      placeholder="Start typing..."
      maxLength={50}
      onChange={onChange}
      value={value}
      ref={reference}
    />
  );
};

TextArea.propTypes = {
  additionalStyles: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  reference: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any })
  ]).isRequired
};

export default TextArea;
