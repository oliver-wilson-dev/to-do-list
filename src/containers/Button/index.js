import { connect } from 'react-redux';
import Button from '../../components/Button';
import incrementCount from '../../state/actions/incrementCount';

const mapDispatchToProps = dispatch => ({
  incrementCount: () => dispatch(incrementCount())
});

export default connect(undefined, mapDispatchToProps)(Button);
