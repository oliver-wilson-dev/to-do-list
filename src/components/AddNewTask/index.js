import React from 'react';
import PropTypes from 'prop-types';
import styles from './index.css';
import Button from '../Button';
import TextArea from '../TextArea';


const AddNewTask = ({ addTask }) => {
  const [description, setDescription] = React.useState('');

  const ref = React.useRef(null);

  const onChange = ({ target: { value: newDescription } }) => {
    setDescription(newDescription);
  };

  const onClick = () => {
    addTask({ description });
    setDescription('');
  };

  return (
    <div className={styles.addNewTask}>
      <h2 className={styles.title}>Add a new task</h2>
      <div className={styles.taskDetails}>
        <TextArea
          additionalStyles={styles.textArea}
          onChange={onChange}
          value={description}
          reference={ref}
        />
        <Button
          additionalStyles={styles.button}
          onClick={onClick}
        >
          <span role="img" aria-label="add a new task">+</span>
        </Button>
      </div>
    </div>
  );
};

AddNewTask.propTypes = {
  addTask: PropTypes.func.isRequired
};

export default AddNewTask;
