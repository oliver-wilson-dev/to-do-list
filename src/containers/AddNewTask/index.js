import { connect } from 'react-redux';
import AddNewTask from '../../components/AddNewTask';
import addTask from '../../state/actions/addTask';

const mapDispatchToProps = dispatch => ({
  addTask: ({ description }) => dispatch(addTask({ description }))
});

export default connect(undefined, mapDispatchToProps)(AddNewTask);
