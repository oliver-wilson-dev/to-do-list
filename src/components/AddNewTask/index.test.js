import React from 'react';
import { shallow } from 'enzyme';
import AddNewTask from './index';
import Button from '../Button';
import TextArea from '../TextArea';

const mockRef = Symbol('test-ref');
jest.spyOn(React, 'useRef').mockImplementation(() => mockRef);

jest.mock('../Button', () => {
  const Button = () => null;

  return Button;
});

jest.mock('../TextArea', () => {
  const TextArea = () => null;

  return TextArea;
});

const mockAddTask = jest.fn();

const defaultProps = { addTask: mockAddTask };

const render = overrideProps => shallow(<AddNewTask {...defaultProps} {...overrideProps} />);

describe('<App/> component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });


  it('should exist', () => {
    expect(render().exists()).toBe(true);
  });

  it('should render correctly', () => {
    expect(render()).toMatchSnapshot();
  });

  it('should render a <Button /> component', () => {
    expect(render().find(Button).exists()).toBe(true);
  });

  it('should render a <TextArea /> component', () => {
    expect(render().find(TextArea).exists()).toBe(true);
  });

  it('should pass the ref to the <TextArea /> component', () => {
    expect(render().find(TextArea).prop('reference')).toEqual(mockRef);
  });

  describe('when rendering the <TextArea />', () => {
    it('should have a default value of an empty string', () => {
      expect(render().find(TextArea).prop('value')).toEqual('');
    });

    describe('when the text area onChange handler is called', () => {
      it('should pass the new description as the value prop to the <TextArea />', () => {
        const newDescription = 'test-new-description';
        const wrapper = render();
        wrapper.find(TextArea).simulate('change', { target: { value: newDescription } });

        expect(wrapper.find(TextArea).prop('value')).toEqual(newDescription);
      });
    });
  });

  describe('when rendering the Button component', () => {
    describe('when the onClick handler is called', () => {
      it('should call addTask with the description', () => {
        const addTask = jest.fn();
        const wrapper = render({ addTask });

        wrapper.find(Button).simulate('click');

        expect(addTask).toHaveBeenCalledWith({ description: '' });
      });

      describe('when the description has been set', () => {
        it('should set the description back to an empty string', () => {
          const addTask = jest.fn();
          const wrapper = render({ addTask });
          const newDescription = 'test-new-description';

          wrapper.find(TextArea).simulate('change', { target: { value: newDescription } });
          wrapper.find(Button).simulate('click');

          expect(wrapper.find(TextArea).prop('value')).toEqual('');
        });
      });
    });
  });
});
