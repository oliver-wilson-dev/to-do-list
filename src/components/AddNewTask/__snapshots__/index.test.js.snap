// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> component should render correctly 1`] = `
<div
  className="addNewTask"
>
  <h2
    className="title"
  >
    Add a new task
  </h2>
  <div
    className="taskDetails"
  >
    <TextArea
      additionalStyles="textArea"
      onChange={[Function]}
      reference={Symbol(test-ref)}
      value=""
    />
    <Button
      additionalStyles="button"
      onClick={[Function]}
    >
      <span
        aria-label="add a new task"
        role="img"
      >
        +
      </span>
    </Button>
  </div>
</div>
`;
